### PR TITLE
Feature/rails form

### PR DESCRIPTION
**Pull request by ChristianOtieno**
Dear reviewer We have implemented the following workings
- Setting up the Back End of our application.
- Building the HTML Form.
- rebuilt the forms with #form_tag
- Rebuilt the forms again with #form_for
- Next, we handled the Editing bit of our users in the database
- Finally, we modified our form view to display a list of the error messages attached to our failed model object should validations fail. 

We hereby welcome your review and comments on our work, thank you!

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
 
 Rails.application.routes.draw do
-  resources :users, only: %i[new create]
+  resources :users, only: %i[new create edit]
```
Comment (Torres-ssf) [inline config/routes.rb]: We also need to include the `:update` route here. We need this route in order to send a POST request to update the user. This request is sent when you click in to update the user.
Screenshot from 2019-12-18 23-15-55

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -13,6 +16,10 @@ def create
     end
   end
 
+  def edit
+    @user = User.find(params[:id])
+  end
+
   private
```
Comment (Torres-ssf) [inline app/controllers/users_controller.rb]: We need to include the update action here. It redirects to edit_user_path if the user was saved (updated) successfully. And also render `:edit `if the user can't be saved(updated)

File: app/views/users/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE html>
+<html>
+<head>
+  <title>HTML FORM</title>
+</head>
+
+<body>
```
Comment (Torres-ssf) [inline app/views/users/edit.html.erb]: We don't need this HTML code here. We already have this code in the application.html.erb file

File: app/views/users/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    <%= f.label :password %>
+    <%= f.password_field :password, placeholder: "Insert your Password"  %>
+
+    <%= f.submit "Submit"%>
+  <% end %> 
+
+</body>
+
+</html>
```
Comment (Torres-ssf) [inline app/views/users/edit.html.erb]: We can delete this since we don't need it

**Comment by ChristianOtieno** [comment]
Dear reviewer, We have implemented the following required changes:
- [x] Re-Addition back of the comment out old form tags.
- [x] Set up of code linter.
- [x] Add update function in the user's routes and add encapsulation to its actions.
- [x] REMOVE redundant markup in both the edit and new embedded ruby files.

File: app/views/users/new.html.erb [hunk 5]
```diff
@@ -1,17 +1,44 @@
...
+  <%= f.label :password %>
+  <%= f.password_field :password, placeholder: "Insert your Password"  %>
+
+  <%= f.submit "Submit"%>
+<% end %> 
+
+
+
+
+<%# <%= form_with(url: "/users", scope: "user", method: "post") do |f| %>
```
Comment (mcrd25) [inline app/views/users/new.html.erb]: This was supposed to be `form_tag` and not `form_with`

**Comment by ChristianOtieno** [comment]
- Changes have been implemented as is required, thank you.

**Comment by mcrd25** [comment]
Fantastic 🙂. Submit the review and someone should get to you soon.

File: app/views/users/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+
+<%= form_for :user do |f| %>
```
Comment (Torres-ssf) [inline app/views/users/edit.html.erb]: Change this line to `<%= form_for(@user, url: user_path) do |f| %>`

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -13,9 +18,45 @@ def create
...
+      if @user.update(user_params)
+        f.html do
+          redirect_to @user, notice: 'User was successfully updated.'
+        end
+      else
+        f.html { render :edit }
+        f.json { render json: @user.errors, status: :unprocessable_entity }
+      end
+    end
+  end
```
Comment (Torres-ssf) [inline app/controllers/users_controller.rb]: And also make the update method look like this.
` def update`
`    @user = User.find(params[:id])`
`    if @user.update(user_params)`
 `     redirect_to new_user_path`
   ` else`
    `  render :edit`
    `end`
 ` end`

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -13,9 +18,45 @@ def create
     end
   end
 
+  # GET /users/1/edit
+  def edit
+    set_user
+  end
```
Comment (Torres-ssf) [inline app/controllers/users_controller.rb]: Change the edit method to
`  def edit`
    `@user = User.find(params[:id])`
  `end`

Comment (ChristianOtieno) [reply app/controllers/users_controller.rb]: Hello, we have the set_user defined already. and it is working fine.

File: app/controllers/users_controller.rb (hunk 7 above)
Comment (ChristianOtieno) [reply app/controllers/users_controller.rb]: Implementation done.

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -26,15 +36,10 @@ def edit
...
-          redirect_to @user, notice: 'User was successfully updated.'
-        end
-      else
-        f.html { render :edit }
-        f.json { render json: @user.errors, status: :unprocessable_entity }
-      end
+    if @user.update(user_params)
+      redirect_to new_user_path
+    else
+      render :edit
```
Comment (samgaco) [inline app/controllers/users_controller.rb]: :+1:

Comment (ChristianOtieno) [reply app/controllers/users_controller.rb]: Thank you!

**Pull request merged**